models/user.py

```diff
@@ -17,6 +16,15 @@ class RoleEnum(Enum):
     STAFF = 3
     ADMIN = 4
 
+    @classmethod
+    def get_values(cls):
+        return [member.value for name, member in cls.__members__.items()]
+
+    @classmethod
+    def has_role(cls, role):
+        role_values = cls.get_values()
+        return role in role_values
```

Comment:
Sure, I could do that.

Reply:
Two quick questions.

1. Do you want me to test for `False` here the same way I did in the integration test (using something like `max(RoleEnum.get_values()) + 1`) or should I hard-code in a value that will likely never get used (e.g., `99`)?

2. When testing for values we expect to return `True`, I figure there are a couple of ways to go about it.  I could loop over each member in `RoleEnum` and assert `RoleEnum.has_value()` on each member's value.  Or I could explicitly create a user with the `empty_test_db` for each of the admin, staff, pm and unauthorized user fixtures and assert `.has_role()` on their role value.  Which would you prefer/think is better?

Or, if you have other suggestions on a better implementation let me know!

Reply:
these are good questions, questions that I have answers too, that I could easily spend an hour talking about. 
The fast answer is:
1. neither
2. We only have 3 roles to test so it probably isn't a bad idea, however exhaustive testing usually isn't scalable and can often be unattainable. But in this case that doesn't apply. I'd be happy with one or all three tests. FYI: unauthorized_user should not have a role. 

OK. Let's talk about question 1 a bit. The goal of testing is to test the behavior and not the implementation. Your solutions for question 1 tests implementation. First it uses a semi-private method that returns the Enums values, and assumes that these values will always be integers, and again the hard-coded input makes another assumption, however I would lean toward the hard-coded input for the test; otherwise your test is dependent on a semi-private method that could become obsolete and removed in the future. Perhaps someone decides that the has_role method doesn't need to depend on the get_values method and removes it and updates the has_role method to:

```
    @staticmethod
    def has_role(role):
        return RoleEnum(role)
```

Ideally all the tests would continue to pass when the implementation is changed, and not the behavior. Although implementing it this way technically causes an implementation change since a value error is now raised when a role is invalid. Although in this form the has_role method doesn't really do much for us now, and can probably be considered to remove it as well. With this method gone we don't need the unit tests anymore. Which raises the question where should test that when we receive input from the client we are validating it accordingly?

Reply:
> FYI: unauthorized_user should not have a role.

The role I was referring to was `PENDING`, which is currently assigned by default, absent a specified role.  The pytest fixture that facilitates testing this right now is called `create_unauthorized_user()`...perhaps this name is misleading?

> OK. Let's talk about question 1 a bit.

Thanks for setting me straight on this one.  I'm basically brand new to writing tests like this, but I'm really enjoying it so far, so I look forward to learning as much as I can. 🤓 

So to test the _behavior_ of the `has_role()` method, I suppose I should be able to pass in any argument of any type into `has_role()` and it should return `False` if that role doesn't exist.

However, moving on to your hypothetical where `has_role()` gets removed:

> With this method gone we don't need the unit tests anymore. Which raises the question: Where should we put tests when we receive input from the client and need to validate it accordingly?

I don't really know the answer to this, but I would guess the test should be directly inside the resource.  If that's the case, though, it seems to me you would mostly be trading away a layer of abstraction in exchange for code that is more stable, but potentially harder to read. Would you agree?

Reply:
The pending role shouldn't be a role. It is slated to go away. Sorry for the confusion on that. In my thought process, I treat it as it is already gone, but I forget that it still exists. 

> I'm basically brand new to writing tests like this, but I'm really enjoying it so far, so I look forward to learning as much as I can.

Very nice. It is a never ending learning process for all of us.

> I don't really know the answer to this, but I would guess the test should be directly inside the resource. If that's the case, though, it seems to me you would mostly be trading away a layer of abstraction in exchange for code that is more stable, but potentially harder to read. Would you agree?

Good assessment, and it would be harder to read if the code is in the resource. However, that is not where it belongs. Where it belongs depends on the framework/architecture being used. In this app there is a place for it and we're slowly moving everything there. My goal this weekend is to update the documentation so that new and existing contributors know how the application is laid out and where stuff belongs. 

There is no reason not to merge this right now. The code is inline with the old architecture which is prevalent in the app right now. If you would like to refactor this in a pairing session and talk more about these topics then I'd be happy to.